app/models/source_resolver.rb

```diff
@@ -0,0 +1,34 @@
+class SourceResolver
+  def self.resolve(new_source)
+    new(new_source).resolve
+  end
+
+  def initialize(new_source)
+    @new_source = new_source
+  end
+
+  def resolve
+    return if @new_source.save
+    adjust_ranks
```

Comment:
Ok, I did think of one situation, but it's pretty edgy - what if two Importers were adding Sources simultaneously? A transaction around this bit of code would prevent them from colliding. Probably.